src/algebra/algebra/spectrum.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright (c) 2021 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import algebra.algebra.basic
+import tactic.noncomm_ring
+/-!
+# Spectrum of an element in an algebra
+This file develops the basic theory of the spectrum of an element of an algebra.
+This theory will serve as the foundation for spectral theory in Banach algebras.
+
+## Main definitions
+
+* `resolvent a`: has type `set R` is the resolvent set of an element `a : A` where
+  `A` is an  `R`-algebra.
+* `spectrum a`: has type `set R` is the spectrum of an element `a : A` where
+  `A` is an  `R`-algebra.
+
+## Main statements
+
+* `unit_left_coset_spectrum`: units in the scalar ring commute (multplication) with the spectrum.
+* `left_add_coset_spectrum`: elements of the scalar ring commute (addition) with the spectrum.
+* `unit_mem_spectrum_mul_iff_swap_mul`: the units (of `R`) in `σ (a*b)` coincide
+  with those in `σ (b*a)`.
+
+## Notations
+
+* `σ a` : `spectrum R a` of `a : A`
+
+## TODO
+
+* Prove the *spectral mapping theorem* for the polynomial functional calculus
+-/
+
+universes u v
+
+namespace spectrum
+
+section defs
+
+variables (R : Type u) {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+-- definition and basic properties
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *resolvent* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+def resolvent (a : A) : set R :=
+{ r : R | is_unit (r • 1 - a) }
+
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *spectrum* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is not a unit of the
+algebra `A`.
+
+The spectrum is simply the complement of the resolvent.  -/
+def spectrum (a : A) : set R :=
+(resolvent R a)ᶜ
+
+end defs
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+local notation `σ` := spectrum R
+
+lemma mem_spectrum_iff {r : R} {a : A} :
+  r ∈ σ a ↔ ¬ is_unit (r • 1 - a) :=
+iff.rfl
+
+lemma not_mem_spectrum_iff {r : R} {a : A} :
+  r ∉ σ a ↔ is_unit (r • 1 - a) :=
+by { apply not_iff_not.mp, simp [set.not_not_mem, mem_spectrum_iff] }
+
+lemma mem_resolvent_of_left_right_inverse {r : R} {a b c : A}
+  (h₁ : (r • 1 - a) * b = 1) (h₂ : c * (r • 1 - a) = 1) :
+  r ∈ resolvent R a :=
+units.is_unit ⟨r•1 - a, b, h₁, by rwa ←left_inv_eq_right_inv h₂ h₁⟩
+
+lemma mem_resolvent_iff {r : R} {a : A} :
+  r ∈ resolvent R a ↔ is_unit (r•1 - a) :=
+iff.rfl
+
+-- products of scalar units and algebra units
+
+lemma is_unit.smul_iff {r : units R} {a : A} :
```

Comment:
Is this true for any group `G`?

```suggestion
lemma is_unit.smul_iff [group G] [mul_action G A] [smul_comm_class G A A] [is_scalar_tower G A A] {r : G} {a : A} :
```

Reply:
This is one example of the namespace issue: right now the full name of this lemma is `spectrum.is_unit.smul_iff`; I think it should just be `is_unit.smul_iff`.

Reply:
yes, this works, I'll address the namespace issue too.